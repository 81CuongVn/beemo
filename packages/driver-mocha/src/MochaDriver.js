/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Driver } from '@beemo/core';

// Success: Writes passed tests to stdout
// Failure: Writes failed tests to stdout
export default class MochaDriver extends Driver {
  bootstrap() {
    this.setMetadata({
      bin: 'mocha',
      configName: 'mocha.opts',
      configOption: '--opts',
      description: 'Unit test files with Mocha.',
      title: 'Mocha',
      useConfigOption: true,
    });
  }

  formatFile(data: Object): string {
    const output = [];

    Object.keys(data).forEach(key => {
      const option =
        key === 'es_staging' || key === 'use_strict' ? `--${key}` : `--${key.replace(/_/g, '-')}`;
      const value = data[key];
      const type = typeof value;

      if (type === 'boolean') {
        output.push(option);
      } else if (type === 'number' || type === 'string') {
        output.push(`${option} ${value}`);
      } else if (Array.isArray(value)) {
        output.push(`${option} ${value.join(',')}`);
      } else if (key === 'reporterOptions') {
        output.push(`${option} ${this.formatReporterOptions(value)}`);
      }
    });

    return output.join('\n');
  }

  formatReporterOptions(options: Object): string {
    return Object.keys(options)
      .map(key => `${key}=${options[key]}`)
      .join(',');
  }
}
