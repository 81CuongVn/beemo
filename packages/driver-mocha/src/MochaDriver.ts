/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Driver } from '@beemo/core';
import { MochaConfig, ReporterOptions } from './types';

// Success: Writes passed tests to stdout
// Failure: Writes failed tests to stdout
export default class MochaDriver extends Driver<MochaConfig> {
  bootstrap() {
    this.setMetadata({
      bin: 'mocha',
      configName: 'mocha.opts',
      configOption: '--opts',
      description: this.tool.msg('app:mochaDriver'),
      title: 'Mocha',
      useConfigOption: true,
      watchOptions: ['-w', '--watch'],
    });
  }

  formatConfig(data: MochaConfig): string {
    const output: string[] = [];

    Object.keys(data).forEach(key => {
      const option =
        key === 'es_staging' || key === 'use_strict' ? `--${key}` : `--${key.replace(/_/gu, '-')}`;
      const value = data[key];
      const type = typeof value;

      if (key === 'reporterOptions') {
        output.push(`${option} ${this.formatReporterOptions(value)}`);
      } else if (type === 'boolean') {
        output.push(option);
      } else if (Array.isArray(value)) {
        if (key === 'globals') {
          output.push(value.join(', '));
        } else {
          value.forEach(v => {
            output.push(`${option} ${v}`);
          });
        }
      } else {
        output.push(`${option} ${value}`);
      }
    });

    return output.join('\n');
  }

  formatReporterOptions(options: ReporterOptions): string {
    return Object.keys(options)
      .map(key => `${key}=${options[key]}`)
      .join(',');
  }
}
