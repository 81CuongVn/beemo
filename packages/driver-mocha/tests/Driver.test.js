import MochaDriver from '../src/MochaDriver';

describe('MochaDriver', () => {
  let driver;

  beforeEach(() => {
    driver = new MochaDriver();
    driver.bootstrap();
  });

  it('sets options from constructor', () => {
    driver = new MochaDriver({
      args: ['--foo', '--bar=1'],
      dependencies: ['babel'],
      env: { DEV: true },
    });

    expect(driver.options).toEqual({
      args: ['--foo', '--bar=1'],
      dependencies: ['babel'],
      env: { DEV: true },
    });
  });

  it('sets correct metadata', () => {
    expect(driver.metadata).toEqual({
      bin: 'mocha',
      configName: 'mocha.opts',
      configOption: '--opts',
      dependencies: [],
      description: 'Unit test files with Mocha.',
      helpOption: '--help',
      title: 'Mocha',
      useConfigOption: true,
    });
  });

  describe('formatFile()', () => {
    it('supports primitives', () => {
      expect(
        driver.formatFile({
          bar: 'abc',
          baz: true,
          foo: 123,
        }),
      ).toMatchSnapshot();
    });

    it('supports arrays', () => {
      expect(
        driver.formatFile({
          ext: ['js', 'jsx'],
        }),
      ).toMatchSnapshot();
    });

    it('handles reporter options', () => {
      expect(
        driver.formatFile({
          reporterOptions: {
            abc: '123',
            def: '456',
          },
        }),
      ).toMatchSnapshot();
    });

    it('handles always underscored options', () => {
      expect(
        driver.formatFile({
          es_staging: true,
          use_strict: true,
        }),
      ).toMatchSnapshot();
    });
  });
});
