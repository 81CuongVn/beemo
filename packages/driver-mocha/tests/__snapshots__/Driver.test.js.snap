// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MochaDriver formatFile() handles always underscored options 1`] = `
"--es_staging
--use_strict"
`;

exports[`MochaDriver formatFile() handles reporter options 1`] = `"--reporterOptions abc=123,def=456"`;

exports[`MochaDriver formatFile() supports arrays 1`] = `"--ext js,jsx"`;

exports[`MochaDriver formatFile() supports primitives 1`] = `
"--bar abc
--baz
--foo 123"
`;
